chore(gatsby): migrate run-sift to TS

pvdz opened the pull request:
Converts run-sift to TS and removes "sift" from the file name.

ascorbic commented on packages/gatsby/src/db/common/query.ts:
```diff
@@ -50,7 +50,7 @@ export function createDbQueriesFromObject(filter: object): Array<DbQuery> {
 
 function createDbQueriesFromObjectNested(
-  filter: object,
+  filter: Record<string, any>,
```
Can this be `Record<string, unknown>`

pvdz replied:
Changing the `any` to `unknown` triggers cascading errors. Are you sure that's the right way?

pvdz replied:
May be related to this:

> packages/gatsby/src/db/common/query.ts:63:57 - error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'Record<string, unknown>'.


I hate that my editor throws different errors from `yarn typecheck` from whatever `yarn format` does. There should be one state :/

ascorbic commented on packages/gatsby/src/redux/run-fast-filters.ts:
```diff
@@ -1,45 +1,60 @@
-
-const { prepareRegex } = require(`../utils/prepare-regex`)
+import { IGatsbyNode } from "./types"
+import { GatsbyGraphQLType } from "../../index"
```
```suggestion
import { GatsbyGraphQLType } from "../.."
```
Don't explicitly import from index

ascorbic commented on packages/gatsby/src/redux/run-fast-filters.ts:
```diff
@@ -1,45 +1,60 @@
+} from "./nodes"
+import { IGraphQLRunnerStats } from "../query/types"
+
+interface IInputQuery {
```
Could you add a comment here explaining why you can't use `Record`, in case someone tries to refactor it

ascorbic commented on packages/gatsby/src/redux/run-fast-filters.ts:
```diff
@@ -1,45 +1,60 @@
+): FilterCacheKey {
   // Note: while `elemMatch` is a special case, in the key it's just `elemMatch`
   // (This function is future proof for elemMatch support, won't receive it yet)
   let f = filter
```
I know this was there already, but I really don't like single char variable names. Letters are free!

pvdz replied:
It was my doing anyways ;)
:shrug: I'll change it

ascorbic commented on packages/gatsby/src/redux/run-fast-filters.ts:
```diff
@@ -295,9 +305,22 @@ function collectBucketForElemMatch(
  *   if `firstOnly` is true
  */
-function runFastFiltersAndSort(args: Object) {
+function runFastFiltersAndSort(args: {
```
Can this be extracted into an interface rather than declaring inline

pvdz commented:
Typing fixed, tests are still broken.

pvdz commented:
Ok. Tests should pass now. Typings were cleaned up.

I thought runFastFilters would accept an array of filters to apply. But apparently, and through the magic of lodash, that's not exactly the case. So I reverted annotating those input/output types as arrays.

Also removed a String case that was accidentally allowing types that shouldn't be allowed. Good thing we have tests :)
